predictions.py

```diff
@@ -0,0 +1,118 @@
+"""
+
+import caffe
+import argparse
+import os
+import numpy as np
+import skimage.exposure as exposure
+import skimage.transform as transform
+import skimage.io as io
+
+
+def standardize(image, size):
+    """
+    Standardize an image to a particular size and rotation
+    :params image: image as numpy array
+    :params size: standard image size as tuple
+    :return: standardized image as numpy array
+    """
+    existing_shape = image.shape
+    if existing_shape[1] > existing_shape[0]:
+        image = transform.rotate(image, 90, resize=True)
+    if image.shape != size:
+        image = transform.resize(image, size)
+    return image
+
+def initialize_model(model_def, model_weights, mean_image):
+    """
+    Initialize a caffe model and data transformer
+    :params model_def: Path to model prototxt file
+    :params model_weights: Path to model weights
+    :params mean_image: Path to mean image file
+    :return net: Net object
+    :return transformer: Transformer object
+    """
+    net = caffe.Net(model_def, model_weights, caffe.TEST)
+
+    mu = np.load(mean_image)
+    mu = mu.mean(1).mean(1)  # average over pixels to obtain the mean (BGR) pixel values
+
+    # create transformer for the input called 'data'
+    transformer = caffe.io.Transformer({'data': net.blobs['data'].data.shape})
+
+    transformer.set_transpose('data', (2,0,1))  # move image channels to outermost dimension
+    transformer.set_mean('data', mu)            # subtract the dataset-mean value in each channel
+    transformer.set_raw_scale('data', 255)      # rescale from [0, 1] to [0, 255]
+    transformer.set_channel_swap('data', (2,1,0))  # swap channels from RGB to BGR
+
+    net.blobs['data'].reshape(1, 3, 227, 227)
+
+    return net, transformer
+
+def augment_image(img):
+    """
+    Augment an image using a combination of lightening, darkening, rotation and mirror images.
+    :params img: Image as numpy array
+    :return: array of augmented images 
+    """
+    augmented_images = []
+    augmented_images.append(np.fliplr(img))
+    for g in [0.45, 0.65, 0.85, 1.25, 1.5, 2]:
+        new_img = exposure.adjust_gamma(img, gamma=g)
+        augmented_images.append(new_img)
+        augmented_images.append(np.fliplr(new_img))
+    new_img = transform.rotate(img, 180)
+    augmented_images.append(new_img)
+    augmented_images.append(np.fliplr(new_img))
+    return np.array(augmented_images)
+
+def simple_predict(img, net, transformer):
+    """
+    Perform simple prediction on a single image by making a forward pass through the network
+    :params img: Image as numpy array
+    :params net: Initialized Net object
+    :params transformer: Initialized Transformer object
+    :return: array of predicted probabilities for each class
+    """
+    net.blobs['data'].data[...] = transformer.preprocess('data', img)
+    p = net.forward()
+    predictions = p['prob'][0].copy()
+    return predictions
+
+def predict_with_augment(img, net, transformer):
```

Comment:
I think what I don't understand here is whether it is just a case of practicality (i.e. additional time and resources), or whether there is something inherently wrong with this approach?

In the end using this approach gave me an extra 2-3% improvement in F1 scores. The prediction time was definitely slower, but still around 0.5 seconds per image.

So is it more just the fact that for bigger or higher volume applications, any accuracy improvements aren't worth the extra time?